3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,271 @@
+     * @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>}
+     */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+      messageContainer => {
+        dev().assert(messageContainer.data,
+          'Received empty new creative message');
+        Object.entries(messageContainer.data).forEach(entry => {
+          const creativeId = entry[0];
+          const extraData = entry[1];
+          if (this.creativeMessageRouters_.hasOwnProperty(creativeId)) {
+            dev().warn(TAG_, 'Duplicate new creative message for ' + creativeId);
+          }
+          this.creativeMessageRouters_[creativeId] =
+            this.creativeMessageRouters_[creativeId] ||
+            new AmpAnalytics3pCreativeMessageRouter(this, creativeId,
+              extraData);
+        });
+      });
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+      messageContainer => {
+        Object.entries(messageContainer.data).forEach(entry => {
+          const creativeId = entry[0];
+          const messages = entry[1];
+          dev().assert(messages.length > 0, 'Received empty events list');
+          dev().assert(this.creativeMessageRouters_[creativeId],
+            'Received event message prior to new creative message.' +
+            ' Discarding.');
+          this.creativeMessageRouters_[creativeId]
+            .receiveEventMessages(messages);
+        });
+        this.flushQueues_();
+      });
+    this.iframeMessagingClient_.sendMessage(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.READY);
+  }
+
+  /**
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Getter to expose iframeMessagingClient to instances of
+   * AmpAnalytics3pCreativeMessageRouter
+   * @returns {!IframeMessagingClient}
+   */
+  getMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+
+  /**
+   * Instructs each AmpAnalytics3pCreativeMessageRouter instances to send its
+   * queued messages to its listener (if it has a listener yet).
+   * @private
+   */
+  flushQueues_() {
+    Object.entries(this.creativeMessageRouters_).forEach(
+      entry => {
+        entry[1].sendQueuedMessagesToListener();
+      });
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative. Also sends repsonse messages from the iframe meant for this
+ * particular creative.
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
```

Comment:
Rather than passing parent, let's just pass the objects within parent needed:  window & iframeMessagingClient.  Better to make dependencies explicit IMO (also allows you to remove getMessagingClient & getWindow functions)